### PR TITLE
feat(agents): knowledge-capture Stop hook and two decant skills

### PR DESCRIPTION
**Pull request by teedole**
Stacked on #84. Base retargets to `main` automatically once that merges, so the
diff here shows only the new work.

## Commit 1, the Stop hook

A hook that compares file edits against knowledge recorded in a session and
prints a reminder on stderr when several files changed and nothing was written
down. Editing AGENTS.md counts as recording, so a session that already captured
its conventions stays quiet.

**This is always on for every contributor, which is the part worth scrutinising.**
Three properties are meant to make that acceptable.

- Offline. Bash and jq only, no network call, no Dosu dependency. The reminder
  names `mcp__dosu__write_knowledge` when that server is listed and falls back to
  AGENTS.md otherwise, so it is useful without Dosu access.
- Silent by default. Read-only sessions and sessions under three edits print
  nothing, and it no-ops entirely without jq.
- Cannot break a session. Always exits 0. The re-entrancy guard on
  `stop_hook_active` keeps it from firing on a stop it caused.

Tested through `bun test` rather than a standalone shell harness, so it runs in
the existing gate instead of needing new plumbing. Eight cases cover the nudge,
small and read-only sessions, recording by either route, the re-entrancy guard, a
missing transcript, and malformed transcript lines. `shellcheck` is clean.

## Commit 2, two skills

Both grounded in this repo, not ported from another codebase.

`new-migration` carries the rule that has actually bitten, which is that a
migration is frozen once committed because a dogfooding archive may already have
run it, so review changes go in a new version. Plus the guarded `db.query`
pattern, the `schema.sql` and `schema-manifest.ts` sync, and the scratch-database
habit that keeps experiments off a live archive.

`new-source-parser` carries the invariant 6 checklist. Synthetic fixtures only,
issues rather than throws on malformed sessions, tests at both parser and ingest
level, and the deliberate `--i-reviewed-the-diff` flag on golden regeneration.

AGENTS.md gains a short pointer under Layout so both skills and the hook are
discoverable without reading `.claude/` directly.

## Deliberately left out

- **No `permissions.allow` entries.** Neither `Skill(*)` nor `mcp__dosu__*`.
  Auto-approving every skill in a public repo that takes outside contributions is
  a wider trust surface than the convenience is worth, so first use prompts.
- **No invariant reminders in the hook.** An earlier draft also nagged about
  `openapi.yaml` drift and frozen migrations based on which paths changed. That
  is scope creep for a hook approved as a knowledge-capture nudge, and the skills
  cover the same ground where it is actually actionable.
- **No `dosu audit` wiring.** AGENTS.md and README are hand-written and current,
  and a single Bun app does not need a generated architecture doc.

## Verification

`bun test` 796 pass 0 fail, `bunx tsc --noEmit` clean, `bunx biome check .`
clean, `shellcheck` clean on the hook.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by teedole** [comment]
Retargeted from #84's branch to `main`. Every workflow in this repo is gated on
`pull_request: branches: [main]`, so a PR based on a feature branch gets no CI at
all, which defeats the point of stacking.

Consequence is that this PR now carries #84's two commits as well, the first two
of the four listed above. Merge #84 first and this diff reduces to just the hook
and the skills on its own. Merge this one alone and it brings both sets, since it
is a strict superset.

File: scripts/claude-hooks/reflect-on-stop.sh [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+set -uo pipefail
+
+INPUT="$(cat)"
+
+# jq does the JSONL walking. Without it the hook is a no-op rather than an error.
+command -v jq >/dev/null 2>&1 || exit 0
+
+# Never re-fire on a stop that this hook itself provoked.
+if [ "$(printf '%s' "$INPUT" | jq -r '.stop_hook_active // false')" = "true" ]; then
```
[the remark this answers is not in the document]
Comment (teedole) [reply scripts/claude-hooks/reflect-on-stop.sh]: Valid, fixed in 1707e57. Both payload reads now discard jq's stderr. Confirmed the failure mode was real rather than theoretical: `printf 'not json at all' | jq -r '.stop_hook_active // false'` prints `jq: parse error: Invalid numeric literal` to stderr, which is exactly the leak you describe.

Also added a test case that would have caught it, covering empty input, non-JSON, a truncated object, and a JSON array, asserting stderr is empty and the exit code is 0 for each. The two jq calls further down already had the redirect, so this brings the payload reads in line with them.

File: scripts/claude-hooks/reflect-on-stop.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  exit 0
+fi
+
+TRANSCRIPT="$(printf '%s' "$INPUT" | jq -r '.transcript_path // empty')"
```
[the remark this answers is not in the document]
Comment (teedole) [reply scripts/claude-hooks/reflect-on-stop.sh]: Valid, fixed in 1707e57. Both payload reads now discard jq's stderr. Confirmed the failure mode was real rather than theoretical: `printf 'not json at all' | jq -r '.stop_hook_active // false'` prints `jq: parse error: Invalid numeric literal` to stderr, which is exactly the leak you describe.

Also added a test case that would have caught it, covering empty input, non-JSON, a truncated object, and a JSON array, asserting stderr is empty and the exit code is 0 for each. The two jq calls further down already had the redirect, so this brings the payload reads in line with them.

File: scripts/claude-hooks/reflect-on-stop.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  ' "$TRANSCRIPT" 2>/dev/null
+}
+
+EDITS="$(count_tools '^(Edit|Write)$')"
+CAPTURED="$(count_tools 'write_knowledge$')"
+EDITS="${EDITS:-0}"
+CAPTURED="${CAPTURED:-0}"
+
+# Writing the repo's own guidance is capture by another route.
+if edited_paths | grep -qE '(AGENTS|CLAUDE)\.md$'; then
```
[the remark this answers is not in the document]
Comment (teedole) [reply scripts/claude-hooks/reflect-on-stop.sh]: Valid, fixed in 1707e57. Now anchored as `(^|/)(AGENTS|CLAUDE)\.md$`.

Worth noting the direction this failed in, since it was the worse of the two options. A match here counts as knowledge being recorded and suppresses the nudge, so `NOT_AGENTS.md` would have silenced the reminder rather than produced a spurious one. Silent-when-it-should-speak is the failure mode this hook can least afford.

Added a test asserting the nudge still fires when the only extra edit is to `NOT_AGENTS.md`.

**Comment by teedole** [comment]
Two things for reviewers beyond the three resolved threads.

**Review fixes.** All three `gemini-code-assist` items were valid and are fixed
in `1707e57`, each with a test that would have caught it. The jq stderr leak was
reproducible rather than theoretical, and the unanchored regex failed in the
dangerous direction, since a match suppresses the nudge and `NOT_AGENTS.md`
would have silenced it.

**Reverted an automated commit.** `e585c16`, pushed to this branch by
`dosubot[bot]` as "docs: Dosu updates for PR #85", deleted the entire
`## Shared knowledge (Dosu)` section from AGENTS.md. 27 lines removed, nothing
added. That section is the substance of the stacked #84, and it contains the
`dosu:mcp` markers that `dosu setup` rewrites in place, so dropping it would also
break that contract on the next setup run.

`16749d6` reverts it. I reverted rather than dropping the commit from history so
the record shows what happened. Flagging it because the bot may push again, and
because a doc agent deleting a section without proposing a replacement looks like
a bug worth a look on its own.
